modules/lkqdBidAdapter.js

```diff
@@ -0,0 +1,217 @@
+      let adHeight = bidRequest.sizes[1];
+      bidRequest.sizes = [[adWidth, adHeight]];
+    }
+
+    for (let j = 0; j < bidRequest.sizes.length; j++) {
+      let size = bidRequest.sizes[j];
+      let playerWidth;
+      let playerHeight;
+      if (size && size.length == 2) {
+        playerWidth = size[0];
+        playerHeight = size[1];
+      } else {
+        utils.logWarn('Warning: Could not determine width/height from the provided adUnit');
+      }
+
+      let sspUrl = ENDPOINT.concat();
+
+      // required parameters
+      sspUrl = _replaceMacro(PID_KEY, bidRequest.params.placementId, PID_DEFAULT, sspUrl);
+      sspUrl = _replaceMacro(SID_KEY, bidRequest.params.siteId, SID_DEFAULT, sspUrl);
+      // optional parameters
+      sspUrl = _replaceMacro(OUTPUT_KEY, bidRequest.params.output, OUTPUT_DEFAULT, sspUrl);
+      sspUrl = _replaceMacro(EXECUTION_KEY, bidRequest.params.execution, EXECUTION_DEFAULT, sspUrl);
+      sspUrl = _replaceMacro(PLACEMENT_KEY, bidRequest.params.placement, PLACEMENT_DEFAULT, sspUrl);
+      sspUrl = _replaceMacro(PLAYINIT_KEY, bidRequest.params.playinit, PLAYINIT_DEFAULT, sspUrl);
+      sspUrl = _replaceMacro(VOLUME_KEY, bidRequest.params.volume, VOLUME_DEFAULT, sspUrl);
+      sspUrl = _replaceMacro(TIMEOUT_KEY, bidRequest.params.timeout, TIMEOUT_DEFAULT, sspUrl);
+      sspUrl = _replaceMacro(WIDTH_KEY, playerWidth, WIDTH_DEFAULT, sspUrl);
+      sspUrl = _replaceMacro(HEIGHT_KEY, playerHeight, HEIGHT_DEFAULT, sspUrl);
+      sspUrl = _replaceMacro(DNT_KEY, bidRequest.params.dnt, DNT_DEFAULT, sspUrl);
+      sspUrl = _replaceMacro(PAGEURL_KEY, bidRequest.params.pageurl, PAGEURL_DEFAULT, sspUrl);
+      sspUrl = _replaceMacro(CONTENTID_KEY, bidRequest.params.contentId, CONTENTID_DEFAULT, sspUrl);
+      sspUrl = _replaceMacro(CONTENTTITLE_KEY, bidRequest.params.contentTitle, CONTENTTITLE_DEFAULT, sspUrl);
+      sspUrl = _replaceMacro(CONTENTLENGTH_KEY, bidRequest.params.contentLength, CONTENTLENGTH_DEFAULT, sspUrl);
+      sspUrl = _replaceMacro(CONTENTURL_KEY, bidRequest.params.contentUrl, CONTENTURL_DEFAULT, sspUrl);
+      // random number to prevent caching
+      sspUrl = sspUrl + '‌&rnd=' + Math.floor(Math.random() * 999999999);
+
+      let sspData = {};
+      sspData.bidId = bidRequest.bidId;
+      sspData.bidWidth = playerWidth;
+      sspData.bidHeight = playerHeight;
+
+      bidRequests.push({
+        method: 'GET',
+        url: sspUrl,
+        data: sspData
+      });
+    }
+  }
+
+  return bidRequests;
+}
+
+function interpretResponse(serverResponse, bidRequest) {
+  let bidResponses = [];
+  if (serverResponse && serverResponse.body) {
+    if (serverResponse.error) {
+      utils.logError('Error: ' + serverResponse.error);
+      return bidResponses;
+    } else {
+      try {
+        let bidResponse = {};
+        if (bidRequest && bidRequest.data && bidRequest.data.bidId && bidRequest.data.bidId !== '') {
+          let sspXml;
+          try {
```

Comment:
As per https://developer.mozilla.org/en-US/docs/Web/API/DOMParser on error handling.
Note that if the parsing process fails, the DOMParser does not throw an exception, but instead returns an error document:

```
<parsererror xmlns="http://www.mozilla.org/newlayout/xml/parsererror.xml">
(error description)
<sourcetext>(a snippet of the source XML)</sourcetext>
</parsererror>
```

Hence the try catch here does not make sense.
This code will still get executed and will result in error when parsing fails
```
bidResponse.creativeId = sspXml.getElementsByTagName('Ad')[0].getAttribute('id');
```

I think you can add something like https://stackoverflow.com/questions/11563554/how-do-i-detect-xml-parsing-errors-when-using-javascripts-domparser-in-a-cross to detect parsing error. This is just a suggestion, there maybe a better solution out there to detect dom parsing errors.

Also add a unit test where xml parsing fails.

Reply:
Have just committed and pushed a fix to address this issue - thank you for catching this, as I was in fact unfamiliar with DOMParser's error handling and my implementation was expecting it to throw an error where it does not. Restructured my logic to account for this and added a new unit test for this case to demonstrate it working as requested.

Reply:
@mrcrawfo I guess you forgot to push the code.

Reply:
You are correct, I just pushed and it is working its way through CI now. Thank you for responding and calling this to my attention.